bindings/python/src/expose-results.hpp

```diff
@@ -31,71 +30,86 @@ exposeResults(pybind11::module_ m)
-    .PROXSUITE_PYTHON_EIGEN_READWRITE(
-      Results<T>,
-      y,
-      "The dual solution associated to the equality constraints.")
-    .PROXSUITE_PYTHON_EIGEN_READWRITE(
-      Results<T>,
-      z,
-      "The dual solution associated to the inequality constraints.")
-    .PROXSUITE_PYTHON_EIGEN_READWRITE(
-      Results<T>,
-      se,
+    // .PROXSUITE_PYTHON_EIGEN_READWRITE(Results<T>, x, "The primal solution.")
+    // .PROXSUITE_PYTHON_EIGEN_READWRITE(
+    //   Results<T>,
+    //   y,
+    //   "The dual solution associated to the equality constraints.")
+    // .PROXSUITE_PYTHON_EIGEN_READWRITE(
+    //   Results<T>,
+    //   z,
+    //   "The dual solution associated to the inequality constraints.")
+    // .PROXSUITE_PYTHON_EIGEN_READWRITE(
+    //   Results<T>,
+    //   se,
+    //   "Optimal shift to the closest feasible problem wrt equality
+    //   constraints.")
+    // .PROXSUITE_PYTHON_EIGEN_READWRITE(Results<T>,
+    //                                   si,
+    //                                   "Optimal shift to the closest feasible
+    //                                   " "problem wrt inequality
+    //                                   constraints.")
+    .def_rw("x", &Results<T>::x, "The primal solution.")
+    .def_rw("y",
+            &Results<T>::y,
+            "The dual solution associated to the equality constraints.")
+    .def_rw("z",
+            &Results<T>::z,
+            "The dual solution associated to the inequality constraints.")
+    .def_rw(
+      "se",
+      &Results<T>::se,
       "Optimal shift to the closest feasible problem wrt equality constraints.")
-    .PROXSUITE_PYTHON_EIGEN_READWRITE(Results<T>,
-                                      si,
-                                      "Pptimal shift to the closest feasible "
-                                      "problem wrt inequality constraints.")
-    .def_readwrite("info", &Results<T>::info)
-    .def(pybind11::self == pybind11::self)
-    .def(pybind11::self != pybind11::self)
-    .def(pybind11::pickle(
-
-      [](const Results<T>& results) {
-        return pybind11::bytes(proxsuite::serialization::saveToString(results));
-      },
-      [](pybind11::bytes& s) {
-        Results<T> results;
-        proxsuite::serialization::loadFromString(results, s);
-        return results;
-      }));
+    .def_rw("si",
+            &Results<T>::si,
+            "Optimal shift to the closest feasible problem wrt inequality "
+            "constraints.")
+    .def_rw("info", &Results<T>::info)
+    .def(nanobind::self == nanobind::self)
+    .def(nanobind::self != nanobind::self)
+    .def("__getstate__",
+         [](const Results<T>& results) {
+           return proxsuite::serialization::saveToString(results);
+         })
+    .def("__setstate__", [](Results<T>& results, const std::string& s) {
+      new (&results) Results<T>{};
+      proxsuite::serialization::loadFromString(results, s);
+    });
```

Comment:
i don't really know how nanobind handles the loading from string. just comparing to before, we don't have a return anymore. is this fine ? i am not 100% sure we are actually testing for serialization in our ci, we should check.

Reply:
So this exists because nanobind collocates the actual object with its binding object. So the memory that represents the C++ object is somewhere inside of the corresponding nanobind object, and this bit of memory is where we are constructing the results object in-place here

Before, pybind just returned objects/pointers everywhere to store them which was really different